### ISSUE TITLE
Allow user defined variability for more parameters

### ISSUE THREAD

**Issue by msevestre** [Open-Systems-Pharmacology/PK-Sim#759]
Dose would be  a great candidate.
Let's list out all potential parameters for which `CanBeVariedInPopulation` should be set to `true`

**Comment by msevestre** [comment]
@stephanschaller
You've mentioned in the past that some parameters should be variable by default. Now is your chance

**Comment by StephanSchaller** [comment]
True, but as I am working in R with all of them set to CanBeVaried (in Pop)= 1, I currently don't have an overview of which parameters are currently available for variation. In my opinion you can allow pretty much all parameters, which are not a formula.

For clinical trial simulation we basically need all (non-formula) event-related parameters:
-  you already mentioned **Dose**
- Dosing Time
- Meals: fraction solid, volume and caloric content, time of event

It seems, all other parameters you could whish for are already available (just checked ;-)). But a nice feature would maybe to allow grouping of parameter variability? (at least I couldn't see anything lime this just now):
I checkt for permeability, and instead of selecting this for every organ, maybe we could just select this for one organ and denn lik the other organs?

@PavelBal, any additional ideas for parameters?

**Comment by Yuri05** [comment]
:no_entry: Dosing Time may not be varied. :no_entry:
With the current implementation of event handling (every dosing is an event) variable dosing times will lead to wrong simulation results

**Comment by msevestre** [comment]
No not for this release whatsoever. I just want to start the discussion.

What do you think it would lead/ not lead to wrong results? 

I see it as an issue because of different time arrays unless we explicitly do not add event start time to the output

**Comment by Yuri05** [comment]
>I see it as an issue because of different time arrays unless we explicitly do not add event start time to the output

We don't add event start times to the output, so the outputs still will have the same length. My initial concern was about estimating of event start times before the simulation. But this should work, because they are estimated for [every simulation run](https://github.com/Open-Systems-Pharmacology/OSPSuite.SimModel/blob/c50568cd40d97c4a1688b925aa4cc128f90def04/src/OSPSuite.SimModel/src/DESolver.cpp#L220)  (and not in `Simulation::Finalize`)

Another concern/problem are PK-values related to the dosing intervals: should we use `Mean(Dosing_time_i)` for calculations?

**Comment by StephanSchaller** [comment]
> should we use Mean(Dosing_time_i) for calculations

It should be calculated relative to the intended dosing time. Variation of dosing time is to simulate delays in applying your dose as planned (few may be earlier, most are later). So it probably looks like a log-normal distribution with a large (> 1) sigma with the intended dosing time at its maximum (e^(µ-sigma^2)).
This may be a problem if intended dosing time is supposed to be t = 0 min as we will get quite a few "Start Times" < 0?

In any case variation of "dosing time" is a nice to have, but I believe if it's too much of an effort, that there are more important features (and anyone really interested can code sth. in R/MATLAB to simulate this)

**Comment by StephanSchaller** [comment]
Also related to a [comment ](https://github.com/Open-Systems-Pharmacology/Forum/issues/180#issuecomment-390599644) in the forum formulation parameters may also be of interest to be varied in a population (particle distribution/size, weibull parameterization)

**Comment by Yuri05** [comment]
>Also related to a comment in the forum formulation parameters may also be of interest to be varied in a population (particle distribution/size, weibull parameterization)

Should not be a problem except for:
:no_entry: particle distribution/size :no_entry:

**Comment by StephanSchaller** [comment]
> My point was actually about calculating PK-Analysis values which relate do dosing start times

Yes, I know, which is why I philosophized about the intended dosing time and how it is reflected in a distribution of varied dosing times ;-)

**Comment by PavelBal** [comment]
To mimic the inter-occasion variability the following parameters might be of interest:

- Parameters related to GI
    -Lumen|LUMEN_SEGMENT|Intestinal transit rate
    -Lumen|LUMEN_SEGMENT|Liquid (start value)
    -Lumen|LUMEN_SEGMENT|Secretion of liquid

- Parameters defining the "fasting" state of the individual. With the current implementation of food intake, parameters "Lumen|Stomach|Liquid/Intestinal transit rate (absolute)/pH" should be calculated from some standard meal and a variable "deltaT = Time - startTime" (see formula "PARAM_MEAL_Lumen_TransportRate_Change" of the meal event). This is probably not so straighforward as an event for "resetting" to fasting state would be necessary.

**Comment by msevestre** [comment]
@Yuri05 
Please list out here the list of parameter that should be changed. I won't implement this in 7.4 otherwise

**Comment by Yuri05** [comment]
>Please list out here the list of parameter that should be changed.

Following parameters should appear under User defined variability:
* `Dose`
* `Fraction endosomal (global)`
* `Thickness (endothelium)`
* `Fraction mucosa`
* `Partition coefficient (water/protein)`
* `Infusion time`
* `Volume of water/body weight`

Following parameters are already available (no change required):
* `Meal energy content`
* `Meal volume`
* `Meal fraction solid`
* `Gallbladder emptying lag time`
<img width="180" alt="aufnahme20" src="https://user-images.githubusercontent.com/25061876/46805690-08098500-cd66-11e8-9e3d-bb70e7e4ca2c.png">

Other parameters discussed above should stay **not available for variation** (no change required)
* _Dosing times_, _Event start times_: as discussed above, this would require some reimplementation and / or additional testing. Not for 7.4 in any case.
* _Particle size distribution_: not possible with the current implementation (Radius/Number of particles in each particles bin are calculated PRIOR to the model creation and cannot be changed after that)
* _Intestinal transit rate_,  _Liquid (start value)_, _Secretion of liquid_: I would check twice (at least) if the variation makes sense. And parameters are currently hidden. Not for 7.4 in any case.

FYI: @sfrechen @Christoph27 @Incei @KatrinCoboeken

**Comment by msevestre** [comment]
Dose might be an issue... It is used to calculate ok analysis and we certainly didn't expect it to be variable. Can we create a separate entry for that an remove it from this list?

**Comment by Yuri05** [comment]
>Can we create a separate entry for that an remove it from this list?

sure

**Comment by Incei** [comment]
I agree with Michael: Assuming we are talking about varying Dose in simulations (not in parameter identification, where I would for sure not vary dose which should be a fixed input parameter where we are relying on):
my understanding of someone wanting to vary a dose would be in a scenario when some individual did not take all the drugs as expected or when the drug got adhesed to a syringe or its tube. But then still I would really think twice before varying it in a pop-sim, e.g. I don't believe that we should calculate a dosing scheme based on simulations where we assume this happens in the first place.

**Comment by msevestre** [comment]
@Yuri05 Actually we do not calculate any Dose dependent PK Analysis for population so varying dose should probably be ok....

**Comment by Incei** [comment]
@msevestre can you explain what you mean with pk-analysis results in a population not being dependent on a dose? :). e.g. a single absolute dose administration for the whole population, wich variability on Dose..

**Comment by msevestre** [comment]
PK Analyses depending directly on the dose are not available for a population simulation. Such as CL. That was my concern because I thought we would then use the wrong dose to calculate them if the dose varied in the population sim. 
But since this is not the case, I don't think that there is any technical issue in making the dose variable in a population simulation

**Comment by Incei** [comment]
Perhaps applied for certain scenarios it might make sense, but i still don't feel comfortable with adding dose variability in a pop sim. 
What about the calculation of pop AUC, Cmax, Half-Life, Cthrough? The pop-CL is not calculated, but is calculated for each individual in the pop, from which usually after exporting out of PK-Sim, a geomean and geoSD clearance is calculated by the user (e.g. for pop comparisons). I would then at least also add the individual Dose info in the 'export pk-analysis to csv' for dose-normalisation.

**Comment by msevestre** [comment]
@Incei As most of the stuff you can do in PK-Sim, this is optional and disabled by default. Varying dose makes sense for certain scenario indeed and there is no reason why we should not support it.

>What about the calculation of pop AUC, Cmax, Half-Life, Cthrough

What about those? This will not change and will remain as expected

Exporting the dose in the pk-analysis export may or may not make sense. It does not feel like the right place to me. It will be part of the population export however if one export the population to csv. Easy to match stuff by IndividualId. At any rate, this should be put in a separate issue.

**Comment by msevestre** [comment]
Actually I am going to revert what I said. We do calculate PK-Parameter normalized to dose for PK-Analyses in population. So varying the dose is much more complicated as expected and this won't be done in this version

**Comment by StephanSchaller** [comment]
This is closed, right?

What Parameters made the list?

I was just trying to vary Kd FcRn (endosome) but could not find it.
I know we have had this discussion before, but can't we just open up all parameters and let the user do the damage? ;-)

**Comment by msevestre** [comment]
I have updated my answer and replace the not with now ;-)

All the one in the list are variable in population.

I think we should look carefully at all parameters that are NOT variable in
population and decide why that is. Some parameters simply cannot be varied
because it will be de facto wrong. Therefore this should not be available
for variation in PKSim.All other parameters, I don't see any problem

**Comment by StephanSchaller** [comment]
Yes, you are right...the relation to physiology is required. Maybe we need another interface then to assess uncertainty.

Let's say, I have a normally distributed error on my measures for Kd FcRn, how do I visualize that impact in the PK curves?
Maybe we need sth like a "Statistics Simulator" with only the "user-defined variability" option on all (or most of) the parameters?


Although, as another example for a missing parameter in the popSim: "fu" is not entirely physiology independent (as we also have the age dependency?).
